input/witnessgen.go

```diff
@@ -2,92 +2,124 @@ package input
 
 import (
 	"fmt"
-
 	"github.com/btcsuite/btcd/txscript"
 	"github.com/btcsuite/btcd/wire"
 )
 
+// WitnessGenerator represents a function which is able to generate the final
+// witness for a particular public key script. Additionally, if required, this
+// function will also return the sigScript for spending nested P2SH witness
+// outputs. This function acts as an abstraction layer, hiding the details of
+// the underlying script.
+type WitnessGenerator func(tx *wire.MsgTx, hc *txscript.TxSigHashes,
+	inputIndex int) (*Script, error)
+
 // WitnessType determines how an output's witness will be generated. The
-// default commitmentTimeLock type will generate a witness that will allow
-// spending of a time-locked transaction enforced by CheckSequenceVerify.
-type WitnessType uint16
+//// default commitmentTimeLock type will generate a witness that will allow
+//// spending of a time-locked transaction enforced by CheckSequenceVerify.
+type WitnessType interface {
+	Type() uint16
```

Comment:
return type `StandardWitnessType`?

Reply:
As discussed it makes sense to not return the stricter type here since an implementing interface would return the value of `CustomType` here.